[envoy] Support TLS in Scalar Envoy chart

kota2and3kan opened the pull request:
## Description

This PR adds TLS support in the Scalar Envoy chart.

In this update, users can configure TLS in both downstream and upstream connections respectively.

```console
[Client] ---(downstream connections)---> [Envoy] ---(upstream connections)---> [ScalarDB/ScalarDL]
```

## Related issues and/or PRs

* https://github.com/scalar-labs/scalar-envoy/pull/28
  * I updated the Scalar Envoy to support TLS both downstream and upstream connections.
* https://github.com/scalar-labs/docs-internal-orchestration/pull/14
  * I updated the document that is related to this PR. You can see the usage of this new feature and the getting started guide.

## Changes made

* Add new values to configure the TLS configurations.
* Mount key/cert file to use TLS connections between client and Envoy in the Deployment manifest.

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

N/A

## Release notes

Support TLS configuration in the Scalar Envoy chart. You can enable TLS in both downstream (client - Envoy) and upstream (Envoy - ScalarDB/ScalarDL) connections.

superbrothers commented on charts/envoy/values.yaml:
```diff
@@ -114,3 +114,21 @@ tolerations: []
+
+tls:
+  # -- TLS configuration between client and Envoy.
+  downstreamTls:
```
nit: Since it is clear from the upper field name that this is a TLS setting, simply `downstream` may be acceptable.

kota2and3kan replied:
Ah, I see. I agree. It's a duplicated field name.
I fixed it in 32f12fd26b75d3dd13d67f6004f1defa24662ee4.

superbrothers commented:
I do not have access to the following documentation, so please give me some basic usage information. How does a user issue a certificate?

- https://github.com/scalar-labs/docs-internal-orchestration/pull/14

If self-signed certificates are used, it is useful for users to have the ability to use cert-manager's Certificate resource in Helm Chart. If you use cert-manager's Certificate resource, cert-manager will automatically renew your certificates.

- https://cert-manager.io/

Helm Charts that support cert-manager often provide this functionality as `certManager.enabled` field.

- https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/values.yaml#L2458-L2468

kota2and3kan commented:
@superbrothers 
Thank you for your review! 🙇‍♀️

> There are typos in the commit message and PR titles in all related PRs. Please fix them.
> 
> > Supprt -> Support

Oh, I overlooked them. Thank you for pointing it out! I fixed them.

> I do not have access to the following documentation, so please give me some basic usage information. How does a user issue a certificate?
> 
> - https://github.com/scalar-labs/docs-internal-orchestration/pull/14

Ah, sorry. We added you with Read privilege to the repository.
Also, I just added you as a reviewer in that PR.
Could you please confirm whether you can see the document PR?

> If self-signed certificates are used, it is useful for users to have the ability to use cert-manager's Certificate resource in Helm Chart. If you use cert-manager's Certificate resource, cert-manager will automatically renew your certificates.
> 
> - https://cert-manager.io/

Ah, I see. We don't force users to use some specific kind of certificate. And, in the getting started guide, we use self-managed CA as a sample.
I think it would be better to provide cert-manager's Certificate resource if users use self-signed certificate so that users can reduce their operations.

> Helm Charts that support cert-manager often provide this functionality as `certManager.enabled` field.
> 
> - https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/values.yaml#L2458-L2468

I will investigate cert-manager and update our helm chart to support it with `certManager.enabled` field in this PR.
Thank you for your suggestion!

superbrothers commented:
> I will investigate cert-manager and update our helm chart to support it with certManager.enabled field in this PR.

If each component is deployed in a different namespace, the configuration can be complex. If deployed in a single namespace, it is simple.

You can also address cert-manager support in a separate PR.

kota2and3kan commented:
@superbrothers  
> > I will investigate cert-manager and update our helm chart to support it with certManager.enabled field in this PR.
> 
> If each component is deployed in a different namespace, the configuration can be complex. If deployed in a single namespace, it is simple.

Thank you for your comment! Basically, we will deploy Envoy and ScalarDB/ScalarDL in the same namespace. So, I think we can make it simple.
However, I think we need some discussions, for example, we have to consider the security requirement in actual systems. I'm discussing it with our business/solution architect team.

> You can also address cert-manager support in a separate PR.

I'm investigating and considering how to integrate cert-manager with our helm chart. But, I think it takes a bit of time.
So, I will do it in another PR.

brfrn169 reviewed (approved):
LGTM! Thank you!

supl reviewed (approved):
LGTM.
Thank you!

feeblefakie reviewed (approved):
LGTM! Thank you!
